File: android/app/src/main/java/com/example/p2pcall/call/CallViewModel.kt
package com.example.p2pcall.call

import android.app.Application
import android.util.Log
import androidx.lifecycle.AndroidViewModel
import androidx.lifecycle.viewModelScope
import com.example.p2pcall.ai.AiCaptionClient
import com.example.p2pcall.ai.AiQaClient
import com.example.p2pcall.ai.AiSummaryClient
import com.example.p2pcall.ai.AudioCaptioner
import com.example.p2pcall.config.Config
import com.example.p2pcall.config.Prefs
import com.example.p2pcall.signaling.SignalingClient
import com.example.p2pcall.webrtc.RtcClient
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import org.json.JSONObject
import org.webrtc.EglBase
import org.webrtc.PeerConnection
import org.webrtc.VideoTrack

/**
 * Room/session state management for Android. Wires the signaling client to the
 * WebRTC client for a 1:1 call and exposes immutable [CallState] plus the local
 * and remote video tracks for rendering. All teardown happens in [cleanup] /
 * [onCleared] so resources are released with the ViewModel lifecycle.
 */
class CallViewModel(app: Application) : AndroidViewModel(app) {

    private val _state = MutableStateFlow(CallState(aiAvailable = Config.aiEnabled))
    val state: StateFlow<CallState> = _state.asStateFlow()

    private val _localTrack = MutableStateFlow<VideoTrack?>(null)
    val localTrack: StateFlow<VideoTrack?> = _localTrack.asStateFlow()

    private val _remoteTrack = MutableStateFlow<VideoTrack?>(null)
    val remoteTrack: StateFlow<VideoTrack?> = _remoteTrack.asStateFlow()

    val eglBase: EglBase = EglBase.create()

    private val prefs = Prefs(app)

    /** Current signaling server URL (persisted override or build-time default). */
    val serverUrl: String get() = prefs.signalingUrl

    private var signaling: SignalingClient? = null
    private var rtc: RtcClient? = null
    private var remotePeerId: String? = null
    private var roomId: String? = null

    // AI captions/summary
    private var captioner: AudioCaptioner? = null
    private var captionClient: AiCaptionClient? = null
    private val transcript = StringBuilder()

    /**
     * Join a room. [serverUrl] overrides the signaling server for this and
     * future sessions (persisted); pass blank to keep the current one. This
     * lets the lobby screen point the app at any private IP / LAN host / URL
     * without rebuilding.
     */
    fun join(room: String, serverUrl: String = "") {
        val trimmed = room.trim()
        if (trimmed.isEmpty()) {
            update { it.copy(status = CallStatus.ERROR, error = "Please enter a room ID.") }
            return
        }

        val url = serverUrl.trim().ifEmpty { prefs.signalingUrl }
        if (url.isEmpty()) {
            update { it.copy(status = CallStatus.ERROR, error = "Please enter a server URL.") }
            return
        }
        // Persist so the chosen server sticks across restarts.
        prefs.signalingUrl = url

        roomId = trimmed
        update { it.copy(status = CallStatus.CONNECTING, roomId = trimmed, error = null) }

        val client = RtcClient(
            context = getApplication(),
            eglBase = eglBase,
            iceServers = Config.iceServers(),
            listener = rtcListener,
        )
        client.initLocalMedia()
        _localTrack.value = client.localVideoTrack
        rtc = client

        val signalingClient = SignalingClient(url, signalingListener)
        signaling = signalingClient
        signalingClient.connect()
    }

    private val signalingListener = object : SignalingClient.Listener {
        override fun onConnected() {
            val room = roomId ?: return
            signaling?.joinRoom(room) { ok, peers, reason ->
                if (!ok) {
                    update {
                        it.copy(
                            status = CallStatus.ERROR,
                            error = if (reason == "room-full") "Room is full (1:1 only)."
                            else "Could not join room."
                        )
                    }
                    return@joinRoom
                }
                if (peers.isNotEmpty()) {
                    // Existing peer present -> we are the initiator.
                    remotePeerId = peers.first()
                    rtc?.createPeerConnection(initiator = true)
                } else {
                    update { it.copy(status = CallStatus.WAITING) }
                }
            }
        }

        override fun onDisconnected() {
            update { it.copy(status = CallStatus.RECONNECTING) }
        }

        override fun onReconnected() {
            // Re-join after a transport reconnect; rebuild the peer connection.
            val room = roomId ?: return
            rtc?.closePeerConnection()
            _remoteTrack.value = null
            signaling?.joinRoom(room) { ok, peers, _ ->
                if (ok && peers.isNotEmpty()) {
                    remotePeerId = peers.first()
                    rtc?.createPeerConnection(initiator = true)
                } else {
                    update { it.copy(status = CallStatus.WAITING) }
                }
            }
        }

        override fun onPeerJoined(peerId: String) {
            // Newcomer is the initiator; we wait for their offer.
            remotePeerId = peerId
            rtc?.createPeerConnection(initiator = false)
        }

        override fun onPeerLeft(peerId: String) {
            if (peerId == remotePeerId) {
                rtc?.closePeerConnection()
                remotePeerId = null
                _remoteTrack.value = null
                update { it.copy(status = CallStatus.WAITING) }
            }
        }

        override fun onSignal(from: String, data: JSONObject) {
            if (remotePeerId == null) {
                remotePeerId = from
                rtc?.createPeerConnection(initiator = false)
            }
            rtc?.handleRemoteSignal(data)
        }
    }

    private val rtcListener = object : RtcClient.Listener {
        override fun onLocalSignal(data: JSONObject) {
            val to = remotePeerId ?: return
            signaling?.sendSignal(to, data)
        }

        override fun onRemoteVideoTrack(track: VideoTrack) {
            _remoteTrack.value = track
            update { it.copy(status = CallStatus.CONNECTED) }
        }

        override fun onConnectionStateChange(state: PeerConnection.PeerConnectionState) {
            viewModelScope.launch {
                when (state) {
                    PeerConnection.PeerConnectionState.CONNECTED ->
                        update { it.copy(status = CallStatus.CONNECTED) }
                    PeerConnection.PeerConnectionState.DISCONNECTED,
                    PeerConnection.PeerConnectionState.FAILED ->
                        update { it.copy(status = CallStatus.RECONNECTING) }
                    else -> Unit
                }
            }
        }
    }

    fun toggleMic() {
        val enabled = !_state.value.micEnabled
        rtc?.setMicEnabled(enabled)
        update { it.copy(micEnabled = enabled) }
    }

    fun toggleCamera() {
        val enabled = !_state.value.cameraEnabled
        rtc?.setCameraEnabled(enabled)
        update { it.copy(cameraEnabled = enabled) }
    }

    /**
     * Toggle live AI captions. Requires RECORD_AUDIO (already granted before
     * joining) and a configured AI server. Opt-in by design.
     */
    fun toggleCaptions() {
        if (!Config.aiEnabled) return
        val enabling = !_state.value.captionsEnabled
        if (enabling) {
            val client = AiCaptionClient(
                wsUrl = Config.aiCaptionsWsUrl,
                onCaption = { line, translation ->
                    transcript.append(line).append('\n')
                    update {
                        it.copy(caption = line, captionTranslation = translation ?: "")
                    }
                },
                onError = { msg -> Log.w("CallViewModel", "captions: $msg") },
            )
            captionClient = client
            val cap = AudioCaptioner(client)
            captioner = cap
            try {
                cap.start()
                // Re-apply any previously chosen translation language.
                client.setLanguage(_state.value.captionLanguage)
                update { it.copy(captionsEnabled = true) }
            } catch (e: SecurityException) {
                update { it.copy(error = "Microphone permission required for captions.") }
            }
        } else {
            captioner?.stop()
            captioner = null
            captionClient = null
            update { it.copy(captionsEnabled = false, caption = "", captionTranslation = "") }
        }
    }

    /** Cycle the live-translation target language: Off → Spanish → French → Hindi → Off. */
    fun cycleCaptionLanguage() {
        if (!Config.aiEnabled) return
        val order = listOf(null, "Spanish", "French", "Hindi")
        val current = order.indexOf(_state.value.captionLanguage)
        val next = order[(current + 1) % order.size]
        captionClient?.setLanguage(next)
        update { it.copy(captionLanguage = next, captionTranslation = "") }
    }

    /** Summarize the accumulated transcript via the FastAPI/Ollama backend. */
    fun requestSummary() {
        if (!Config.aiEnabled || transcript.isBlank()) return
        val text = transcript.toString()
        update { it.copy(summarizing = true, error = null) }
        viewModelScope.launch {
            try {
                val result = withContext(Dispatchers.IO) {
                    AiSummaryClient(Config.aiSummaryUrl).summarize(text)
                }
                update {
                    it.copy(
                        summarizing = false,
                        summary = result.summary,
                        actionItems = result.actionItems,
                    )
                }
            } catch (e: Exception) {
                update { it.copy(summarizing = false, error = "Summary failed: ${e.message}") }
            }
        }
    }

    fun dismissSummary() {
        update { it.copy(summary = null, actionItems = emptyList()) }
    }

    /** Ask a question grounded in the live transcript (FastAPI/Ollama /ask). */
    fun ask(question: String) {
        if (!Config.aiEnabled || question.isBlank() || transcript.isBlank()) return
        val text = transcript.toString()
        update { it.copy(asking = true, error = null) }
        viewModelScope.launch {
            try {
                val answer = withContext(Dispatchers.IO) {
                    AiQaClient(Config.aiAskUrl).ask(text, question)
                }
                update { it.copy(asking = false, answer = answer) }
            } catch (e: Exception) {
                update { it.copy(asking = false, error = "Ask failed: ${e.message}") }
            }
        }
    }

    fun dismissAnswer() {
        update { it.copy(answer = null) }
    }

    fun hangup() {
        cleanup()
        update { CallState(status = CallStatus.ENDED, aiAvailable = Config.aiEnabled) }
    }

    private fun cleanup() {
        captioner?.stop()
        captioner = null
        captionClient = null
        transcript.setLength(0)
        roomId?.let { signaling?.leaveRoom(it) }
        signaling?.disconnect()
        signaling = null
        rtc?.dispose()
        rtc = null
        remotePeerId = null
        roomId = null
        _localTrack.value = null
        _remoteTrack.value = null
    }

    private fun update(transform: (CallState) -> CallState) {
        _state.value = transform(_state.value)
    }

    override fun onCleared() {
        super.onCleared()
        try {
            cleanup()
            eglBase.release()
        } catch (e: Exception) {
            Log.e("CallViewModel", "Error in onCleared", e)
        }
    }
}

File: ios/P2PCall/Call/CallViewModel.swift
import Foundation
import Combine
import WebRTC

/// Room/session state management for iOS. Wires the signaling client to the
/// WebRTC client for a 1:1 call and publishes immutable `CallState` plus the
/// local and remote video tracks for rendering. All teardown happens in
/// `cleanup()` so resources are released deterministically. The iOS counterpart
/// to Android's `CallViewModel`.
@MainActor
final class CallViewModel: ObservableObject {

    @Published private(set) var state = CallState(aiAvailable: Config.aiEnabled)
    @Published private(set) var localTrack: RTCVideoTrack?
    @Published private(set) var remoteTrack: RTCVideoTrack?

    private var signaling: SignalingClient?
    private var rtc: RtcClient?
    private var remotePeerId: String?
    private var roomId: String?

    // AI captions/summary
    private var captioner: AudioCaptioner?
    private var captionClient: AiCaptionClient?
    private var transcript = ""

    // MARK: - Join / signaling

    func join(_ room: String) {
        let trimmed = room.trimmingCharacters(in: .whitespacesAndNewlines)
        if trimmed.isEmpty {
            update { $0.status = .error; $0.error = "Please enter a room ID." }
            return
        }
        roomId = trimmed
        update { $0.status = .connecting; $0.roomId = trimmed; $0.error = nil }

        let client = RtcClient(iceServers: Config.iceServers(), listener: rtcListener())
        client.initLocalMedia()
        localTrack = client.localVideoTrack
        rtc = client

        let signalingClient = SignalingClient(url: Config.signalingUrl, listener: signalingListener())
        signaling = signalingClient
        signalingClient.connect()
    }

    private func signalingListener() -> SignalingClient.Listener {
        var l = SignalingClient.Listener()

        l.onConnected = { [weak self] in
            Task { @MainActor in
                guard let self = self, let room = self.roomId else { return }
                self.signaling?.joinRoom(room) { [weak self] ok, peers, reason in
                    Task { @MainActor in
                        guard let self = self else { return }
                        if !ok {
                            self.update {
                                $0.status = .error
                                $0.error = reason == "room-full"
                                    ? "Room is full (1:1 only)."
                                    : "Could not join room."
                            }
                            return
                        }
                        if let first = peers.first {
                            // Existing peer present -> we are the initiator.
                            self.remotePeerId = first
                            self.rtc?.createPeerConnection(initiator: true)
                        } else {
                            self.update { $0.status = .waiting }
                        }
                    }
                }
            }
        }

        l.onDisconnected = { [weak self] in
            Task { @MainActor in self?.update { $0.status = .reconnecting } }
        }

        l.onReconnected = { [weak self] in
            Task { @MainActor in
                guard let self = self, let room = self.roomId else { return }
                // Re-join after a transport reconnect; rebuild the peer connection.
                self.rtc?.closePeerConnection()
                self.remoteTrack = nil
                self.signaling?.joinRoom(room) { [weak self] ok, peers, _ in
                    Task { @MainActor in
                        guard let self = self else { return }
                        if ok, let first = peers.first {
                            self.remotePeerId = first
                            self.rtc?.createPeerConnection(initiator: true)
                        } else {
                            self.update { $0.status = .waiting }
                        }
                    }
                }
            }
        }

        l.onPeerJoined = { [weak self] peerId in
            Task { @MainActor in
                guard let self = self else { return }
                // Newcomer is the initiator; we wait for their offer.
                self.remotePeerId = peerId
                self.rtc?.createPeerConnection(initiator: false)
            }
        }

        l.onPeerLeft = { [weak self] peerId in
            Task { @MainActor in
                guard let self = self else { return }
                if peerId == self.remotePeerId {
                    self.rtc?.closePeerConnection()
                    self.remotePeerId = nil
                    self.remoteTrack = nil
                    self.update { $0.status = .waiting }
                }
            }
        }

        l.onSignal = { [weak self] from, data in
            Task { @MainActor in
                guard let self = self else { return }
                if self.remotePeerId == nil {
                    self.remotePeerId = from
                    self.rtc?.createPeerConnection(initiator: false)
                }
                self.rtc?.handleRemoteSignal(data)
            }
        }

        return l
    }

    private func rtcListener() -> RtcClient.Listener {
        var l = RtcClient.Listener()

        l.onLocalSignal = { [weak self] data in
            Task { @MainActor in
                guard let self = self, let to = self.remotePeerId else { return }
                self.signaling?.sendSignal(to: to, data: data)
            }
        }

        l.onRemoteVideoTrack = { [weak self] track in
            Task { @MainActor in
                guard let self = self else { return }
                self.remoteTrack = track
                self.update { $0.status = .connected }
            }
        }

        l.onConnectionStateChange = { [weak self] newState in
            Task { @MainActor in
                guard let self = self else { return }
                switch newState {
                case .connected:
                    self.update { $0.status = .connected }
                case .disconnected, .failed:
                    self.update { $0.status = .reconnecting }
                default:
                    break
                }
            }
        }

        return l
    }

    // MARK: - Media toggles

    func toggleMic() {
        let enabled = !state.micEnabled
        rtc?.setMicEnabled(enabled)
        update { $0.micEnabled = enabled }
    }

    func toggleCamera() {
        let enabled = !state.cameraEnabled
        rtc?.setCameraEnabled(enabled)
        update { $0.cameraEnabled = enabled }
    }

    // MARK: - AI features

    /// Toggle live AI captions. Requires microphone permission (granted before
    /// joining) and a configured AI server. Opt-in by design.
    func toggleCaptions() {
        guard Config.aiEnabled else { return }
        let enabling = !state.captionsEnabled
        if enabling {
            let client = AiCaptionClient(
                wsUrl: Config.aiCaptionsWsUrl,
                onCaption: { [weak self] line, translation in
                    Task { @MainActor in
                        guard let self = self else { return }
                        self.transcript += line + "\n"
                        self.update {
                            $0.caption = line
                            $0.captionTranslation = translation ?? ""
                        }
                    }
                },
                onError: { msg in print("CallViewModel captions: \(msg)") }
            )
            captionClient = client
            let cap = AudioCaptioner(client: client)
            captioner = cap
            do {
                try cap.start()
                // Re-apply any previously chosen translation language.
                client.setLanguage(state.captionLanguage)
                update { $0.captionsEnabled = true }
            } catch {
                update { $0.error = "Microphone permission required for captions." }
            }
        } else {
            captioner?.stop()
            captioner = nil
            captionClient = nil
            update {
                $0.captionsEnabled = false
                $0.caption = ""
                $0.captionTranslation = ""
            }
        }
    }

    /// Cycle the live-translation target language: Off → Spanish → French → Hindi → Off.
    func cycleCaptionLanguage() {
        guard Config.aiEnabled else { return }
        let order: [String?] = [nil, "Spanish", "French", "Hindi"]
        let current = order.firstIndex(where: { $0 == state.captionLanguage }) ?? 0
        let next = order[(current + 1) % order.count]
        captionClient?.setLanguage(next)
        update { $0.captionLanguage = next; $0.captionTranslation = "" }
    }

    /// Summarize the accumulated transcript via the FastAPI/Ollama backend.
    func requestSummary() {
        guard Config.aiEnabled, !transcript.isEmpty else { return }
        let text = transcript
        update { $0.summarizing = true; $0.error = nil }
        Task {
            do {
                let result = try await AiSummaryClient(summaryUrl: Config.aiSummaryUrl).summarize(text)
                update {
                    $0.summarizing = false
                    $0.summary = result.summary
                    $0.actionItems = result.actionItems
                }
            } catch {
                update { $0.summarizing = false; $0.error = "Summary failed: \(error.localizedDescription)" }
            }
        }
    }

    func dismissSummary() {
        update { $0.summary = nil; $0.actionItems = [] }
    }

    /// Ask a question grounded in the live transcript (FastAPI/Ollama /ask).
    func ask(_ question: String) {
        guard Config.aiEnabled, !question.isEmpty, !transcript.isEmpty else { return }
        let text = transcript
        update { $0.asking = true; $0.error = nil }
        Task {
            do {
                let answer = try await AiQaClient(askUrl: Config.aiAskUrl).ask(transcript: text, question: question)
                update { $0.asking = false; $0.answer = answer }
            } catch {
                update { $0.asking = false; $0.error = "Ask failed: \(error.localizedDescription)" }
            }
        }
    }

    func dismissAnswer() {
        update { $0.answer = nil }
    }

    // MARK: - Teardown

    func hangup() {
        cleanup()
        state = CallState(status: .ended, aiAvailable: Config.aiEnabled)
    }

    private func cleanup() {
        captioner?.stop()
        captioner = nil
        captionClient = nil
        transcript = ""
        if let room = roomId { signaling?.leaveRoom(room) }
        signaling?.disconnect()
        signaling = nil
        rtc?.dispose()
        rtc = nil
        remotePeerId = nil
        roomId = nil
        localTrack = nil
        remoteTrack = nil
    }

    private func update(_ transform: (inout CallState) -> Void) {
        var copy = state
        transform(&copy)
        state = copy
    }

    deinit {
        // ViewModel is @MainActor; cleanup() touches main-actor state, so hop on.
        let captioner = self.captioner
        let signaling = self.signaling
        let rtc = self.rtc
        captioner?.stop()
        signaling?.disconnect()
        rtc?.dispose()
    }
}
